client/examples/perf.rs

```diff
@@ -0,0 +1,466 @@
+        }
+
+        Ok(measurements)
+    }
+
+    /// Run the connections test and output the result
+    pub async fn test(args: &Options, conn_cmd: &ConnectionsCommand) -> anyhow::Result<()> {
+        let measurements = measure_connections(args, conn_cmd).await?;
+
+        if let Some(path) = &args.csv {
+            let file = std::fs::File::create(path)?;
+            measurements.save(file)?;
+        } else {
+            measurements.save(std::io::stdout())?;
+        }
+
+        Ok(())
+    }
+}
+
+mod load_test {
+    use super::*;
+
+    /// Testable request
+    #[derive(Clone, Copy, Debug)]
+    enum Request {
+        FindNode,
+        Neighbours(u32),
+        Ping,
+    }
+
+    /// Singular test case consisting of repeating a request a set number of times
+    #[derive(Clone, Debug)]
+    struct Task {
+        request: Request,
+        count: u32,
+    }
+
+    /// Testing plan for a connection
+    #[derive(Clone, Debug)]
+    struct TaskPlan {
+        tasks: Vec<Task>,
+    }
+
+    impl TaskPlan {
+        fn count_requests(&self) -> u32 {
+            self.tasks.iter().map(|task| task.count).sum()
+        }
+    }
+
+    #[derive(Debug)]
+    struct DataPoint {
+        connections: u32,
+        rate_limit: u32,
+        rate: f32,
+    }
+
+    #[derive(Debug)]
+    struct Measurements {
+        data: Vec<DataPoint>,
+    }
+
+    impl Measurements {
+        fn new() -> Self {
+            Measurements { data: Vec::new() }
+        }
+
+        fn push(&mut self, connections: u32, rate_limit: u32, rate: f32) {
+            self.data.push(DataPoint {
+                connections,
+                rate_limit,
+                rate,
+            });
+        }
+
+        fn save(&self, writer: impl std::io::Write) -> anyhow::Result<()> {
+            let mut w = csv::Writer::from_writer(writer);
+
+            w.write_record(&["connections", "rate-limit", "rate"])?;
+            for data_point in &self.data {
+                w.write_record(&[
+                    &data_point.connections.to_string(),
+                    &data_point.rate_limit.to_string(),
+                    &data_point.rate.to_string(),
+                ])?;
+            }
+
+            Ok(())
+        }
+    }
+
+    /// Block execution to achieve a consistent delay between calls
+    async fn limit_rate(start: &mut Instant, delay: Option<f32>) {
+        if let Some(secs) = delay {
+            let duration = Duration::from_secs_f32(secs);
+
+            loop {
+                let now = Instant::now();
+                if now > *start + duration {
+                    let overshoot = now - (*start + duration);
+                    *start = now - overshoot;
+
+                    break;
+                }
+
+                let diff = (*start + duration - now) / 3;
+                tokio::time::sleep(diff).await;
+            }
+        }
+    }
+
+    /// Execute the plan on one connection
+    async fn test1(
```

Comment:
Why don't name function properly as in comment?
Consider doing that in other places where relevant